runtime/compiler/z/codegen/J9TreeEvaluator.cpp

```diff
@@ -3396,6 +3198,455 @@ J9::Z::TreeEvaluator::ArrayCopyBNDCHKEvaluator(TR::Node * node, TR::CodeGenerato
+      }
+   else
+      {
+      dataSnippet = static_cast<TR::S390WritableDataSnippet *>(cg->CreateConstant(node, dataBlock, 1 << (int) (log2(sizeof(J9JITWatchedInstanceFieldData)-1)+1), true));
+      }
+   return dataSnippet;
+   }
+
+void
+J9::Z::TreeEvaluator::readWriteBarrierHelperForFieldWatch(TR::CodeGenerator *cg, TR::Node *node, TR::Snippet *dataSnippet, bool isWrite)
+   {
+   bool isResolved = !node->getSymbolReference()->isUnresolved();
+   TR::LabelSymbol *mergePointLabel = generateLabelSymbol(cg);
+   TR::LabelSymbol *fieldReportLabel = generateLabelSymbol(cg);
+
+   TR::Register *fieldClassReg;
+   TR::Register *fieldClassFlags = cg->allocateRegister();
+   bool opCodeIsIndirect = node->getOpCode().isIndirect();
+
+   if (opCodeIsIndirect)
+      {
+      TR::Register *objReg = cg->evaluate(node->getFirstChild());
+      // Load the class of the instance object into fieldClassReg.
+      fieldClassReg = cg->allocateRegister();
+      TR::TreeEvaluator::genLoadForObjectHeadersMasked(cg, node, fieldClassReg, generateS390MemoryReference(objReg, static_cast<int32_t>(TR::Compiler->om.offsetOfObjectVftField()), cg), NULL);
+      }
+   else
+      {
+      if (isResolved)
+         {
+         fieldClassReg = cg->allocateRegister();
+         J9JITWatchedStaticFieldData *staticFieldDataBlock = reinterpret_cast<J9JITWatchedStaticFieldData *>((static_cast<TR::S390WritableDataSnippet *>(dataSnippet))->getRawData());
+         J9Class *fieldClass = staticFieldDataBlock->fieldClass;
+         TR_ASSERT_FATAL(fieldClass != NULL, "A valid J9Class must be provided for direct rdbar/wrtbar opcodes %p\n", node);
+         generateRILInstruction(cg, TR::InstOpCode::LARL, node, fieldClassReg, static_cast<void *>(fieldClass));
+         }
+      else
+         {
+         fieldClassReg = cg->evaluate(isWrite ? TR::TreeEvaluator::getDirectWrtbarJ9ClassNode(node) : node->getFirstChild());
+         }
+      }
+   // First load the class flags into a register.
+   generateRXInstruction(cg, TR::InstOpCode::getLoadOpCode(), node, fieldClassFlags, generateS390MemoryReference(fieldClassReg, cg->comp()->fej9()->getOffsetOfClassFlags(), cg));
+   // Then test the bit to test with the relevant flag to check if fieldwatch is enabled.
+   generateRIInstruction(cg, TR::InstOpCode::TMLL, node, fieldClassFlags, J9ClassHasWatchedFields);
+   // If Condition Code from above test is not 0, then we branch to OOL (instructions) to report the fieldwatch event. Otherwise fall through to mergePointLabel.
+   generateS390BranchInstruction(cg, TR::InstOpCode::BRC, TR::InstOpCode::COND_BRNZ, node, fieldReportLabel);
+
+   // Generate instructions to call a VM Helper and report the fieldwatch event. Also generates an instruction to
+   // branch back to mainline (mergePointLabel).
+   generateReportFieldAccessOutlinedInstructions(node, fieldReportLabel, mergePointLabel, dataSnippet, isWrite, cg);
```

Comment:
Is this function really necessary? I only see one place in which we call it which is here.

Reply:
Yeah I think we should leave it in here for now. It keeps the code a little more organized IMO. More importably, when we have support for `TR_OutlinedInstructionsGenerator` in the z/codegen as well we can easily refactor this code out into the common codegen with some minor cleanup. Here's the x86 codegen version of this call:
https://github.com/eclipse/openj9/blob/6bf7e0de73f5e1ce5daac1d709d07fca97733900/runtime/compiler/x/codegen/J9TreeEvaluator.cpp#L12705-L12708

As is obvious, this code can be refactored once we have implemented the necessary support.

Reply:
Ok, thanks for the explanation. I understand now.